### PR TITLE
Implement BlockCipher trait

### PR DESCRIPTION
**Pull request by Xanewok**
This implements the `BlockCipher` trait from the https://github.com/RustCrypto/traits interface, so that we can use the CNG as the underlying block cipher for other cryptographic primitives in Rust. The primary motivation is to use it for the EAX mode (authenticated encryption) in the [`eax`](https://crates.io/crates/eax) crate (currently moved under the [RustCrypto/AEADs](https://github.com/RustCrypto/AEADs/tree/master/eax) umbrella repository).

This is a WIP PR outlining the proposed changes. As it was done in my asymmetric-related PRs, this also exposes the supported algorithm to a compile-time enumeration - the reason for that is some RustCrypto traits require block size at the compile-time and so we need to also know here which algorithm we're using at the compile-time as well.

This introduces basically the same mechanism for `SymmetricAlgorithmId` and `Algorithm` trait - maybe it'd make sense to introduce a generic trait for usage with asymmetric/symmetric/hash/other? domains? Something like:
```rust
pub trait Algorithm<Id> {
    // e.g. e.g. `Some(Aes)` or `None` for the runtime-dispatched enum like `SymmetricAlgorithmId`
    // To be used in compile-time
    const ID: Option<Id>;
    // Returns a type-erased ID value, e.g. `SymmetricAlgorithmId`
    // To be used at run-time
    fn id(&self) -> Id;

    // Maybe this might be useful (either for internal or public usage) as well?
    fn name(&self) -> &str { self.id().as_str() }
}

// ...

pub struct SymmetricKey<A: Algorithm<Id = SymmetricAlgorithmId>> { /* ... */ }
// ...now that I just wrote it, I'm not sure that bounds for algorithm with specified associated types are
// possible/stable - will need to investigate that
```

@emgre what do you think?

### CODE REVIEW

**Comment by Xanewok** [comment]
Updated the implementation, it should be now easier to use the block ciphers that have different key lengths in generic contexts. At the moment I tested and implemented support for AES 128/192/256.

I'd appreciate if you could take a moment and let me know what you think.

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]: This should be moved to the `#[cfg(test)]` block.

File: Cargo.toml
Comment (emgre) [inline Cargo.toml]: We should an execution in the pipeline with this trait on/off, similar to what we did with `zeroize`.

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]: I don't get how the `BlockCipher` will decide what chaining mode to use?

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]: Indeed, that would be nice to have. An assert on the size of the provided array will be required.

**Review by emgre — COMMENTED** [review]
I'm open to this PR. The one global trait (`block_cipher` stuff) with different implementations was the original goal of this crate.

File: src/symmetric.rs
Comment (Xanewok) [inline src/symmetric.rs]: @emgre This is where we specify using ECB. As I understand it, `BlockCipher` are meant to be used only as a block cipher primitives, which can be later wrapped by different mode structs. Because of this, I'm specifying ECB and passing `None` IV to have a 'transparent/plain' AES encryption as `BlockCipher` implementation

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: See https://github.com/emgre/win-crypto-ng/pull/31/files#r503954069

Comment (Xanewok) [reply src/symmetric.rs]: I forgot about that we fact that we can cast `SymmetricAlgorithmKey` to Key - that's where using other chaining modes than ECB will not be possible...

One solution is to encode chaining mode alongside `KeyBits` as a generic parameter and only then `impl BlockCipher for Key<Aes/* for now */, Ecb, {128,192,256}>` but I'm worried about encoding more and more details in the compile-time context.

The other is to either put a friendly warning that only ECB keys support `BlockCipher` and let it blow up or `panic!` ourselves if the chaining mode for the key used for `encrypt_block`/`decrypt_block` is not in ECB mode.

Do you have any idea and/or preference how we should handle this?

File: src/symmetric.rs
Comment (Xanewok) [inline src/symmetric.rs]: In theory this can be implemented for any `T: typenum::Unsized` - do you want me to do that? I didn't do that for brevity in `impl BlockCipher for super::Key` (we don't have to specify other `typenum` bounds because we don't need `PartialDiv` in `type KeySize = ...` like we do in `NewBlockCipher`) but we could reduce duplication here.

File: src/symmetric.rs
Comment (emgre) [reply src/symmetric.rs]: Oh ok, so `BlockCipher` or whatever other crate on top of it will do the actual chaining. I get it.

File: src/symmetric.rs
Comment (emgre) [reply src/symmetric.rs]: I suspect that the other algorithms (such as RC2) will require a lot more of `typenum`, so we should probably implement it for `T: typenum::Unsized` here.

File: src/symmetric.rs
Comment (emgre) [reply src/symmetric.rs]: I'm not convinced by the extra type, especially since it would only be used to check that it's ECB. Seems a bit complicated for not much. Perhaps the `panic!` is the best option. We could panic in the `encrypt_block` / `decrypt_block` or straight in the `TryFrom`.

Comment (Xanewok) [reply src/symmetric.rs]: I'm not 100% sure either is a good location, unfortunately:
- `*_block` operations will be called somewhat often and so it'd be great to check if we can panic only once
- If we intend to have a general compile-time known handle to a symmetric key (which it should be perfectly valid and even preferred to have a different chaining mode than ECB) then we shouldn't panic in `TryFrom` only to have a useful trait implemented.

Maybe having a helper struct similar to [`std::path::Display`](https://doc.rust-lang.org/std/path/struct.Display.html) would be a correct choice? If we couple it with a fallible conversion, then we can only do the check once

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: Done in ce9e11c45547212877629dea8f4007bd26501216

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: Done in 900a51b1277dfa8457e7e61d497edfa008239da6

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: I implemented that approach in a9c9aaa, let me know what you think.

File: Cargo.toml
Comment (Xanewok) [reply Cargo.toml]: Done in 923af62

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: Do you want to block this PR on having this implemented using a new in-place API? I'd prefer to implement that as an improvement in a separate PR, if you don't mind

**Comment by Xanewok** [comment]
@emgre this should be more or less ready for an another review round!

File: src/symmetric.rs
Comment (emgre) [reply src/symmetric.rs]: Yeah, let's do it in another PR. I created #32 to not forget about it.

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]: Why Windows, just why... 😓

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]:
```suggestion
        /// Returns `Ok` if and only if the underlying key is set to ECB chaining mode, see
```

File: src/symmetric.rs
Comment (emgre) [inline src/symmetric.rs]: Before merging this PR, should we implement `BlockCipher` for other algorithms than AES? My understanding of the code is that we can create a `BlockCipherKey` from any algorithm with `try_into_block_cipher`, but only AES actually implements the `BlockCipher` trait, right?

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: Done in https://github.com/emgre/win-crypto-ng/pull/31/commits/ff58ce47ad650dc07021721b93fd42d9075f06c9.

File: src/symmetric.rs
Comment (Xanewok) [reply src/symmetric.rs]: That's right. I implemented `(New)BlockCipher` for remaining block ciphers in https://github.com/emgre/win-crypto-ng/pull/31/commits/2b54a6d6e5261259894e62374e060d1367716570 and https://github.com/emgre/win-crypto-ng/pull/31/commits/364361d74a95dafbb6e29295311e9a455f03d1d8.

Comment (Xanewok) [reply src/symmetric.rs]: This way, you can instantiate the `BlockCipher`-compatible key via either
1. `BlockCipherKey::<Algo, Bits>::new(...)` (using `block_cipher::NewBlockCipher`) or
2. `Key::try_into_block_cipher`, `Key` can be obtained via `TryFrom` from a dynamic `SymmetricAlgorithmKey`

Because of the design of `BlockCipher`, you're generally encouraged AIUI to be generic over ciphers implementing *both* `BlockCipher` and `NewBlockCipher`, and so if we wouldn't do that we'd lose quite a bit of compatibility across the crypto ecosystem.

In general, it's good to have a strictly typed (keysize-wise) constructor for block ciphers but the fact that this adds a 3rd public symmetric key type might be slightly confusing, rather than just having a compile-time `Key` and dynamic `SymmetricAlgorithmKey`. The users should ideally be encouraged to use those two and only use `Key::try_into_block_cipher` if they wish to have a `BlockCipher`-capable key.

I think the benefit outweights the possible confusion in this case, and we could return to this later at some point, possibly conflating `BlockCipherKey` with `Key` (by exposing the chaining mode as a type parameter).

**Review by emgre — APPROVED** [review]

**Pull request merged**
